File: kotlin/src/main/kotlin/com/edgeveda/sdk/ResourceMonitor.kt
package com.edgeveda.sdk

import android.app.ActivityManager
import android.content.Context
import android.os.Debug
import android.util.Log
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Monitors memory resource usage (RSS - Resident Set Size).
 *
 * Tracks the app's memory footprint for budget enforcement and telemetry.
 * Uses Android [Runtime] for Java heap and [Debug] for native heap measurement.
 *
 * Thread-safe via [Mutex] (Kotlin equivalent of Swift actor isolation).
 *
 * Example:
 * ```kotlin
 * val monitor = ResourceMonitor()
 * val currentMemory = monitor.currentRssMb()
 * val peakMemory = monitor.peakRssMb()
 * ```
 */
class ResourceMonitor(
    private val maxSamples: Int = 100
) {
    private val mutex = Mutex()
    private val samples = mutableListOf<Double>()
    private var _peakRssMb: Double = 0.0

    companion object {
        private const val TAG = "EdgeVeda.ResourceMonitor"
    }

    /**
     * Current RSS (Resident Set Size) in megabytes.
     *
     * Triggers a fresh memory sample before returning.
     */
    suspend fun currentRssMb(): Double = mutex.withLock {
        updateMemoryUsage()
        samples.lastOrNull() ?: 0.0
    }

    /** Peak RSS observed since monitoring started. */
    suspend fun peakRssMb(): Double = mutex.withLock { _peakRssMb }

    /** Average RSS over the sample window. */
    suspend fun averageRssMb(): Double = mutex.withLock {
        if (samples.isEmpty()) 0.0 else samples.sum() / samples.size
    }

    /** Number of samples collected. */
    suspend fun sampleCount(): Int = mutex.withLock { samples.size }

    /**
     * Manually trigger a memory usage update.
     *
     * Memory is automatically sampled when accessing [currentRssMb],
     * but this method allows explicit sampling for telemetry purposes.
     */
    suspend fun sample() = mutex.withLock {
        updateMemoryUsage()
    }

    /** Reset all samples and peak tracking. */
    suspend fun reset() = mutex.withLock {
        samples.clear()
        _peakRssMb = 0.0
    }

    // -----------------------------------------------------------------------
    // Private
    // -----------------------------------------------------------------------

    /** Must be called under [mutex] lock. */
    private fun updateMemoryUsage() {
        val rss = getResidentSetSize()
        samples.add(rss)

        if (rss > _peakRssMb) {
            _peakRssMb = rss
        }

        // Keep sliding window
        if (samples.size > maxSamples) {
            samples.removeFirst()
        }
    }

    /**
     * Get the current approximate resident set size in megabytes.
     *
     * On Android there is no direct equivalent of macOS `mach_task_basic_info`.
     * We approximate RSS by summing:
     * - Java heap usage: [Runtime.totalMemory] - [Runtime.freeMemory]
     * - Native heap usage: [Debug.getNativeHeapAllocatedSize]
     *
     * For a more precise PSS reading, use [getMemoryInfoFromContext] with
     * an [ActivityManager] (requires a [Context]).
     */
    private fun getResidentSetSize(): Double {
        return try {
            val runtime = Runtime.getRuntime()
            val javaHeapBytes = runtime.totalMemory() - runtime.freeMemory()
            val nativeHeapBytes = Debug.getNativeHeapAllocatedSize()
            val totalBytes = javaHeapBytes + nativeHeapBytes
            totalBytes.toDouble() / (1024.0 * 1024.0)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get memory info: ${e.message}")
            0.0
        }
    }
}

File: swift/Sources/EdgeVeda/ResourceMonitor.swift
import Foundation

#if os(iOS) || os(macOS)
import os.log
#endif

/// Monitors memory resource usage (RSS - Resident Set Size).
///
/// Tracks the app's memory footprint for budget enforcement and telemetry.
/// RSS represents the actual physical memory used by the process.
///
/// Example:
/// ```swift
/// let monitor = ResourceMonitor()
/// let currentMemory = await monitor.currentRssMb
/// let peakMemory = await monitor.peakRssMb
/// ```
@available(iOS 15.0, macOS 12.0, *)
actor ResourceMonitor {
    private var samples: [Double] = []
    private let maxSamples = 100
    private var _peakRssMb: Double = 0.0
    
    #if os(iOS) || os(macOS)
    private let logger = Logger(subsystem: "com.edgeveda.sdk", category: "ResourceMonitor")
    #endif
    
    // MARK: - Initialization
    
    init() {
        // Take initial sample
        updateMemoryUsage()
    }
    
    // MARK: - Public Properties
    
    /// Current RSS (Resident Set Size) in megabytes.
    var currentRssMb: Double {
        updateMemoryUsage()
        return samples.last ?? 0.0
    }
    
    /// Peak RSS observed since monitoring started.
    var peakRssMb: Double {
        return _peakRssMb
    }
    
    /// Average RSS over the sample window.
    var averageRssMb: Double {
        guard !samples.isEmpty else { return 0.0 }
        return samples.reduce(0.0, +) / Double(samples.count)
    }
    
    /// Number of samples collected.
    var sampleCount: Int {
        return samples.count
    }
    
    // MARK: - Public Methods
    
    /// Manually trigger a memory usage update.
    ///
    /// Memory is automatically sampled when accessing currentRssMb,
    /// but this method allows explicit sampling for telemetry purposes.
    func sample() {
        updateMemoryUsage()
    }
    
    /// Reset all samples and peak tracking.
    func reset() {
        samples.removeAll()
        _peakRssMb = 0.0
    }
    
    // MARK: - Private Methods
    
    private func updateMemoryUsage() {
        let rss = getResidentSetSize()
        samples.append(rss)
        
        // Update peak
        if rss > _peakRssMb {
            _peakRssMb = rss
        }
        
        // Keep sliding window
        if samples.count > maxSamples {
            samples.removeFirst()
        }
    }
    
    /// Get the current resident set size in megabytes.
    private func getResidentSetSize() -> Double {
        var info = mach_task_basic_info()
        var count = mach_msg_type_number_t(MemoryLayout<mach_task_basic_info>.size) / 4
        
        let result = withUnsafeMutablePointer(to: &info) {
            $0.withMemoryRebound(to: integer_t.self, capacity: 1) {
                task_info(
                    mach_task_self_,
                    task_flavor_t(MACH_TASK_BASIC_INFO),
                    $0,
                    &count
                )
            }
        }
        
        guard result == KERN_SUCCESS else {
            #if os(iOS) || os(macOS)
            logger.error("Failed to get memory info: \(result)")
            #else
            print("Failed to get memory info: \(result)")
            #endif
            return 0.0
        }
        
        // Convert bytes to megabytes
        let rssBytes = Double(info.resident_size)
        let rssMb = rssBytes / (1024.0 * 1024.0)
        
        return rssMb
    }
}

